streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java

```diff
@@ -168,29 +171,59 @@ public void run() {
                 ProducerRecord<byte[], byte[]> record =
                         new ProducerRecord<>("data", stringSerde.serializer().serialize("", key), intSerde.serializer().serialize("", value));
 
-                producer.send(record, new Callback() {
-                    @Override
-                    public void onCompletion(final RecordMetadata metadata, final Exception exception) {
-                        if (exception != null) {
-                            exception.printStackTrace();
-                            Exit.exit(1);
-                        }
-                    }
-                });
-
+                producer.send(record, new TestCallback(record, needRetry));
 
                 numRecordsProduced++;
                 allData.get(key).add(value);
                 if (numRecordsProduced % 100 == 0)
                     System.out.println(numRecordsProduced + " records produced");
                 Utils.sleep(2);
+            }
+        }
+        producer.flush();
 
+        int remainingRetries = 5;
+        while (!needRetry.isEmpty()) {
+            final List<ProducerRecord<byte[], byte[]>> needRetry2 = new ArrayList<>();
+            for (final ProducerRecord<byte[], byte[]> record : needRetry) {
+                producer.send(record, new TestCallback(record, needRetry2));
+            }
+            producer.flush();
+            needRetry = needRetry2;
+
+            if (--remainingRetries == 0) {
```

Comment:
Note that we are effectively only retrying 4 times with this logic, since if we happen to successfully send out all the record on the fifth retry (i.e. `needRetry` is not empty) we are still going to throw error here.

So more strictly we should check ` --remainingRetries == 0 && !needRetry.isEmpty()`.

Reply:
Ack. Good catch.